packages/realm-server/migrations/1728561353454_create-user-billing.js

```diff
@@ -0,0 +1,202 @@
+exports.up = (pgm) => {
+  pgm.createTable('users', {
+    id: {
+      type: 'uuid',
+      primaryKey: true,
+      default: pgm.func('gen_random_uuid()'),
+    },
+    matrix_user_id: { type: 'varchar', notNull: true, unique: true },
+    stripe_customer_id: { type: 'varchar' },
+    created_at: {
+      type: 'timestamp',
+      notNull: true,
+      default: pgm.func('current_timestamp'),
+    },
+  });
+
+  pgm.createTable('plans', {
+    id: {
+      type: 'uuid',
+      primaryKey: true,
+      default: pgm.func('gen_random_uuid()'),
+    },
+    name: { type: 'varchar', notNull: true },
+    monthly_price: { type: 'numeric', notNull: true },
+    credits_included: { type: 'integer', notNull: true },
+    created_at: {
+      type: 'timestamp',
+      notNull: true,
+      default: pgm.func('current_timestamp'),
+    },
+  });
+
+  pgm.createTable('subscription_cycles', {
+    id: {
+      type: 'uuid',
+      primaryKey: true,
+      default: pgm.func('gen_random_uuid()'),
+    },
+    subscription_id: { type: 'uuid' },
+    period_start: { type: 'timestamp', notNull: true },
+    period_end: { type: 'timestamp', notNull: true },
+  });
+
+  pgm.createType('subscription_status', ['active', 'canceled', 'expired']);
+
+  pgm.createTable('subscriptions', {
+    id: {
+      type: 'uuid',
+      primaryKey: true,
+      default: pgm.func('gen_random_uuid()'),
+    },
+    user_id: { type: 'uuid', references: 'users(id)' },
+    plan_id: { type: 'uuid', references: 'plans(id)' },
+    started_at: { type: 'timestamp', notNull: true },
+    ended_at: { type: 'timestamp', notNull: true },
+    status: { type: 'subscription_status', notNull: true },
+    stripe_subscription_id: { type: 'varchar', notNull: true },
+    current_billing_cycle_id: {
+      type: 'uuid',
+      references: 'subscription_cycles(id)',
+    },
+  });
+
+  pgm.addConstraint(
+    'subscription_cycles',
+    'subscription_cycles_subscription_id_fkey',
+    {
+      foreignKeys: {
+        columns: 'subscription_id',
+        references: 'subscriptions(id)',
+      },
+    },
+  );
+
+  pgm.createTable('ai_actions', {
+    id: {
+      type: 'uuid',
+      primaryKey: true,
+      default: pgm.func('gen_random_uuid()'),
+    },
+    user_id: { type: 'uuid', references: 'users(id)' },
+    cost_in_usd: { type: 'numeric', notNull: true },
+    created_at: {
+      type: 'timestamp',
+      notNull: true,
+      default: pgm.func('current_timestamp'),
+    },
+  });
+
+  pgm.createType('balance_change_type', [
+    'plan_allowance',
+    'extra_credit_balance',
```

Comment:
Do we need another entry here for "used"?

Reply:
And perhaps one for "expired", to be used when the credits you received for plan_allowance expire at the end of the cycle if unused?

Reply:
> Do we need another entry here for "used"

So the idea is that we will sum the debits (used credits) in `credit_balance_changes` (will be renamed to credit_balance_events) for the current subscription period and we will get the amount of used credits in this period.  There are 2 types of used credits in `credit_balance_changes`:  one from the monthly allowance and one from "extra credits". I believe this approach doesn't need the "used" entries. 

> And perhaps one for "expired"

Similarly as above, I think there's no need for that as well. If we filter by subscription cycle period and sum the usage from the mentioned table, we can simply compare it with the plan's limit, then think there's no need to have an idea of "expired credits" (because every month the sum aggregation will calculate the total consumption only in this period). The "extra" credits is different, these don't expire and we can get the available amount by summing up credits and debits for this credit type. 

So the gist of this is that we don't actually credit (make an entry in balance events) for the monthly plan allowance but filter by period and sum the usage and compare it with the plan's limit. If user is at the limit then we'll also check if there is any "extra credits" available to spend.

In my mind this would work but eager to see if you see any potential issues with this!

Reply:
If a user has 100 credits per month from their plan and purchases a top up of 100 credits in month one, and then uses 190 credits in month one and then tries to use 120 credits in month two, How do we know whether they have enough credits available to do so? In other words, it's easy to get the total extra credits, but how are we tracking how many are spent and how many are available?

Reply:
Just had a call with @lukemelia and talked through this, we decided in favor of adding new types to flag "used" and "expired" credits. The change is also that beginning of every period, we will make an entry for new plan allowance with the value of the plan's limit, and make an entry to expire the plan allowance from the previous month so that it doesn't roll over into the new month. This will make things more explicit and help with pro-rationing when users upgrade to a bigger plan.

Reply:
@lukemelia following up on that, and looking at the designs, where we want to communicate current consumption in two **separate** amounts (plan allowance, extra credits), then I think we need to do this a bit more granularly–by introducing not only `used`, but `plan allowance used` and `extra credit used`.

<img width="297" alt="image" src="https://github.com/user-attachments/assets/f2e63465-e2a8-465c-8d97-89d576263151">

<img width="711" alt="image" src="https://github.com/user-attachments/assets/d0e9be95-c7f5-4e25-a86c-69ae1086bbd8">

Reply:
Stripe sends a webhook when renewal happens. In our case this will be every month, starting from when the user signed up for the plan. When this happens, we will insert a new subscription cycle, and add new credit for this month's allowance (and expire the remaining credit from previous month). So no cron needed here.